include/device_vector.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <vector>
+
+#include <quda_api.h>
+#include <malloc_quda.h>
+#include <util_quda.h>
+#include <complex_quda.h>
+
+/**
+   @file device_vector.h
+   Helper functionality for handling a small piece of device memory.
+ */
+
+namespace quda
+{
+  /**
+    @brief This class `device_vector` provides a light weighted wrapper for data on device.
+  */
+
+  template <class real> class device_vector
+  {
+    real *_device_data = nullptr;
+
+    size_t _size;
+
+  public:
+    /**
+       @brief constructor. The device memory is allocated according to the size of the host_vector,
+       and the content of the host_vector is copied to the device memory
+       @param[in] host_vector the host vector
+    */
+    device_vector(std::vector<real> &host_vector) : _size(host_vector.size())
+    {
+      size_t bytes = _size * sizeof(real);
+      if (bytes > 0) {
+        _device_data = reinterpret_cast<real *>(device_malloc(bytes));
```

Comment:
This should probably be `pool_device_malloc` ?

Reply:
https://github.com/lattice/quda/pull/1061/commits/377701f828ca406411764326773a4dcfdefd8b74